### ISSUE TITLE
dice metric incorrect?

### ISSUE THREAD

**Issue by justinblaber** [fastai/fastai#2267]
```
def dice(input:Tensor, targs:Tensor, iou:bool=False, eps:float=1e-8)->Rank0Tensor:
    "Dice coefficient metric for binary target. If iou=True, returns iou metric, classic for segmentation problems."
    n = targs.shape[0]
    input = input.argmax(dim=1).view(n,-1)
    targs = targs.view(n,-1)
    intersect = (input * targs).sum().float()
    union = (input+targs).sum().float()
    if not iou: return (2. * intersect / union if union > 0 else union.new([1.]).squeeze())
    else: return (intersect / (union-intersect+eps) if union > 0 else union.new([1.]).squeeze())
```

This appears to be computing dice for the entire batch. So if the batch size is greater than 1 and the `union` is zero for some indices in the batch, but not others in the `batch`, it will compute dice incorrectly. 

Another issue: if the training set is not divisible by the batch size, then I assume the last batch will be smaller than the rest, so the average value computed at the end for the validation set (a `mean()` of `mean()`'s) will not be correct?

**Comment by justinblaber** [comment]
Something like this should work (although I haven't tested it yet):

```
def dice(input:Tensor, targs:Tensor)->Rank0Tensor:
    "Dice coefficient metric for binary target."
    
    # Threshold targs
    n = targs.shape[0]
    input = input.argmax(dim=1).view(n,-1)
    targs = targs.view(n,-1) 
    
    # Compute dice
    intersect = (input * targs).sum(dim=1).float()
    union = (input + targs).sum(dim=1).float()    
    dice = 2. * intersect / union
    
    # Replace zero union values with 1
    dice[union == 0.] = 1
    
    # Return mean
    return dice.mean()
```

**Comment by sgugger** [comment]
The means are adjusted for various batch sizes by the library, so this part isn't a problem. Happy to look at a PR fixing this!

**Comment by justinblaber** [comment]
Pull request here: #2278

**Comment by sgugger** [comment]
Closing since the PR was merged.
